### ISSUE TITLE
In MasterDetailPage if you set the Detail to NavigationPage(someContentPage) and navigate to another content page, then click the back button from the nav bar, breakpoint from OnBackButtonPressed will not fire.

### ISSUE THREAD

**Issue by kwiqniss** [xamarin/Xamarin.Forms#3184]
### Description

If you use a MasterDetail pattern with Detail = new NavigationPage(myContentPage),
then navigate from that content page to another content page, you'll see a back button in the navigation bar if HasBackButton is true.
When the back button is pressed, OnBackButtonPressed should be raised on the content page. However the event is not fired. [UPDATE, the event fires, but does not trigger breakpoint] [UPDATE, the event fires in UWP. In Android, the event doesn't fire and the breakpoint is not hit]

### Steps to Reproduce
1. Create MasterPage and ContentPage to use as detail page.
2. In your ContentPage, navigate to another content page using Navigation.PushAsync. 
3. In the ContentPage you navigate to, override OnBackButtonPressed. Set a breakpoint here
3. Create MasterDetail set Detail = new NavigationPage(myDetailPage). 
4. When an item is selected from the detail page, navigate to the second contentPage
5. from the content page, click the back button located in the navigation bar

### Expected Behavior
after step 5, it is expected that the breakpoint in the ContentPage's OnBackButtonPressed would be hit.
[UPDATE] We also expect Debug.WriteLine statements to show in Debug Output when running Debug build.
### Actual Behavior
On Windows
after step 5, we observe that the breakpoint in the detail page's OnBackButtonPressed handler is not triggered, however, we do notice the output in Debug output
On iOS and Android, the event does not fire.
### Basic Information

- Version with issue: 3.1.0.583944 
- Last known good version: unsure.
- IDE: Visual Studio or Visual Studio for Mac
- Platform Target Frameworks: 
issue occurs on iOS, Android, and UWP desktop and mobile devices, simulators, and emulators.

my visual studio about:
Microsoft Visual Studio Enterprise 2017 
Version 15.7.4
VisualStudio.15.Release/15.7.4+27703.2035
Microsoft .NET Framework
Version 4.7.03056

Installed Version: Enterprise

Visual C++ 2017   00369-60000-00001-AA728
Microsoft Visual C++ 2017

Application Insights Tools for Visual Studio Package   8.12.10405.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.0.40601.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.7.31476
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.60419.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.0.40608.0
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.3.3000.5
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   15.0.40617.0
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.3.3000.5
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

GitHub.VisualStudio   2.5.3.2888
A Visual Studio Extension that brings the GitHub Flow into Visual Studio.

ILSpy.AddIn   1.0
Integration of the ILSpy Decompiler into Visual Studio.

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

JavaScript UWP Project System   2.0
JavaScript UWP Project System

Merq   1.1.19-rc (a4ffc1b)
Command Bus, Event Stream and Async Manager for Visual Studio extensions.

Microsoft Azure HDInsight Azure Node   2.3.3000.5
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.3.3000.5
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   2.2
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.3.3000.5
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10420.2

Microsoft Continuous Delivery Tools for Visual Studio   0.3
Simplifying the configuration of continuous build integration and continuous build delivery from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

Mono Debugging for Visual Studio   4.10.5-pre (ab58725)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Redgate SQL Change Automation   3.0.18169.138
Extend DevOps processes to your SQL Server databases and safely automate database deployments.

Visit https://www.red-gate.com/sca/productpage for more information.

Copyright (C) 2011 Red Gate Software Ltd. All rights reserved.

This software contains components from Component Owl.
SQL Server is a registered trademark of Microsoft Corporation.
Visual Studio is a registered trademark of Microsoft Corporation.

SQL Change Automation contains code from the following open source software:

NuGet https://www.nuget.org/
SQL LocalDB Wrapper https://github.com/martincostello/sqllocaldb
Autofac https://autofac.org/
Json.NET https://json.net/
MahApps.Metro http://mahapps.com/
SemVer https://github.com/maxhauser/semver
Log4Net http://logging.apache.org/log4net/
StringTemplate https://github.com/antlr/stringtemplate4
Extended WPF Toolkit https://wpftoolkit.codeplex.com/
Code InfoBox VSX http://www.codeproject.com/Articles/55196/Code-InfoBox-Visual-Studio-Extension-VSX
OctoPack https://github.com/OctopusDeploy/OctoPack
SQLite https://sqlite.org/

This product contains icons from http://www.visualpharm.com distributed under a free backlink license.

For license details or other notices relating to the above software, please see NOTICE.TXT and EULA.rtf in the SQL Change Automation application folder.
    

Redgate SQL Prompt   9.1.16.5356
Write, format, and refactor SQL effortlessly

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

Snapshot Debugging Extension   1.0
Snapshot Debugging Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.61804.210
Microsoft SQL Server Data Tools

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   15.7.20419.2003
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.8.3-beta6-62923-07. Commit Hash: 7aafab561e449da50712e16c9e81742b8e7a2969
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual C++ for Cross Platform Mobile Development (Android)   15.0.27625.00
Visual C++ for Cross Platform Mobile Development (Android)

Visual C++ for Cross Platform Mobile Development (iOS)   15.0.27520.00
Visual C++ for Cross Platform Mobile Development (iOS)

Visual F# Tools 10.1 for F# 4.1   15.7.0.0.  Commit Hash: 173513e369ffb7a1c4d5dccf83696d9aac2ab2d0.
Microsoft Visual F# Tools 10.1 for F# 4.1

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Apache Cordova   15.123.7408.1
Visual Studio Tools for Apache Cordova

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

Visual Studio Tools for Universal Windows Apps   15.0.27703.2035
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Windows Machine Learning Generator Extension   1.0
Windows Machine Learning Visual Studio Extension Detailed Info

Windows Template Studio   2.2.18159.1
Windows Template Studio quickly builds a UWP app, using a wizard-based UI to turn your needs into a foundation of Windows 10 patterns and best practices.

Xamarin   4.10.10.2 (35a01d8dc)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.12.1 (f3257e429)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin.Android SDK   8.3.3.2 (HEAD/dffc59120)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   11.12.0.4 (64fece5)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

**Comment by kwiqniss** [comment]
I'm attaching a project. In this project, there's a Master Detail Page. The Master page has navigation options to the Items page and to the about page.
By default, Items page is selected.
In the Items page, click an Item.
The ItemDetailPage will open.
You'll notice a back button in the nav bar.
In the ItemDetailPage, I overrode OnBackButtonPressed.
Typically, in content page wrapped in NavigationPage, when this back button is pressed, you'd see that OnBackButtonPressed will fire. In this example, the event does not fire.

I ran this sample on Windows, using Visual Studio 2017. I launched the UWP project and ran as Debug, x64 on local machine.
[MasterDetailSample.zip](https://github.com/xamarin/Xamarin.Forms/files/2151233/MasterDetailSample.zip)

**Comment by kwiqniss** [comment]
I proved this was firing, only the breakpoint isn't being hit.
I updated my sample so ItemsPage.xaml.cs fires using async
`private async void ItemSelected(object sender, SelectedItemChangedEventArgs e)
        {
            await Navigation.PushAsync(new ItemDetailPage((Item)e.SelectedItem));
        }`

I updated ItemDetailPage.xaml.cs to include Debug.WriteLine statements to see if event fired.

`public ItemDetailPage (Item item)
		{
			InitializeComponent ();
            this.item = item;
            this.Title = $"{item.Id} detail";
            BindingContext = item;
            Debug.WriteLine($"Make Sure Debug Writeline is working so we can be sure it works if we do not see output from {nameof(OnBackButtonPressed)}");
        }

        protected override bool OnBackButtonPressed()
        {
            Debug.WriteLine($"{nameof(OnBackButtonPressed)} fired successfully!");
            return base.OnBackButtonPressed();
        }`
In the output, as expected, I see 

> Make Sure Debug Writeline is working so we can be sure it works if we do not see output from OnBackButtonPressed
OnBackButtonPressed fired successfully!

However, what's not expected is that my breakpoint at that location was skipped.

**Comment by kwiqniss** [comment]
I added a call to Debbugger.Break(). When I encounter that, my debugger stops and shows message like:

> MasterDetailSample.pdb not loaded

Module Information
Version: 1.00.0.0
Original Location: C:\src\MasterDetailSample\MasterDetailSample\MasterDetailSample\MasterDetailSample.UWP\bin\x86\Debug\AppX\MasterDetailSample.dll

**Comment by kwiqniss** [comment]
I get the same results in iOS that I get in Android.
Here's debugger output:

> Resolved pending breakpoint at 'ItemDetailPage.xaml.cs:32,1' to bool MasterDetailSample.Views.ItemDetailPage.OnBackButtonPressed () [0x0001c].
Resolved pending breakpoint at 'ItemDetailPage.xaml.cs:31,1' to bool MasterDetailSample.Views.ItemDetailPage.OnBackButtonPressed () [0x00007].
[0:] Make Sure Debug Writeline is working so we can be sure it works if we do not see output from OnBackButtonPressed

**Comment by hartez** [comment]
I tried running your example, and the Debug statement shows up in the console when I press the back button on Android.

Keep in mind that `OnBackButtonPressed` only applies to the hardware back button; it does not apply to the back button in the navigation bar: https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.navigationpage.onbackbuttonpressed?view=xamarin-forms

**Comment by kwiqniss** [comment]
It's an inconsistent behavior. In any other content page, this event fires when the back button in nav bar is pressed. It doesn't fire on Android and iOS when done from MasterDetailsPage, but does fire if you aren't using MasterDetailsPage. It's also inconsistent that it fires on Windows, but not android and iOS when done from MasterDetailsPage.

**Comment by hartez** [comment]
If a NavigationPage is the Detail of a MasterDetailPage, the navigation bar back button does not trigger OnBackButtonPressed on iOS or Android. The hardware back button triggers OnBackButtonPressed on Android.

In a standalone NavigationPage (no MasterDetailPage), the navigation bar back button does not trigger OnBackButtonPressed on iOS or Android. The hardware back button triggers OnBackButtonPressed on Android. 

On UWP, the navigation bar back button is effectively the same button as the "hardware" back button; both are the built-in system navigation button, they just show up in different places in desktop vs. tablet mode. In either mode, it follows the same rules as the Android hardware button - OnBackButtonPressed is triggered.

In other words, 
On iOS - no "hardware" button, so the method is never called
On UWP - all "hardware" buttons, so the method is always called
On Android - called for the "hardware" button, never called for the "software" button

I'm attaching a simple repro to demonstrate this. If I'm missing something, let me know.

[BackButtonPressed Repro.zip](https://github.com/xamarin/Xamarin.Forms/files/2172217/BackButtonPressed.Repro.zip)

**Comment by kwiqniss** [comment]
You'll see in page and content page that there is no specific mention of this only working for hardware button. The issue is occurring in a content page. If you do OnBackButtonPressed in any other content page where the page is wrapped in a NavigationPage, clicking back button in nav bar will fire OnBackButtonPressed, unless that ContentPage is set to the Details of a MasterDetailPage. So, it's inconsistent behavior for NavigationPage(new MyContentPage). 
Also, we have inconsistencies between platforms for this behavior. When you do have NavPage(ContentPage), then OnBackButtonPressed fires for Windows (although, breakpoint isn't hit when debugging) but it doesn't fire at all on Android or iOS.
This is a bug. I believe you read the wrong documentation.
Please reactivate this.
I'm a Microsoft SEE. This is a valid issue.

**Comment by hartez** [comment]
> If you do OnBackButtonPressed in any other content page where the page is wrapped in a NavigationPage, clicking back button in nav bar will fire OnBackButtonPressed

Run the reproduction project I posted (BackButtonPressed Repro.zip) on Android:

1. Select "Navigation Page". This sets the main page of the application to a NavigationPage which wraps a ContentPage containing a button.
2. Tap the button on the ContentPage. This pushes a ContentPage onto the stack which overrides OnBackButtonPressed.
3. Tap the hardware back button at the bottom of the device. The console will display "BackButtonPressedOverridePage OnBackButtonPressed".
4. Tap the button on the ContentPage again.
5. Tap the "back" button in the navigation bar at the top of the screen. The console will not display the "BackButtonPressedOverridePage OnBackButtonPressed" message.

I believe this repro project is illustrating the situation you are describing (a ContentPage wrapped in a NavigationPage), and the behavior is as claimed in the documentation - the hardware back button will invoke OnBackButtonPressed, the navigation bar back button will not. Is there a detail about the situation I'm missing?

**Comment by kwiqniss** [comment]
This is odd. I watched Xam U instructor use this for Content Page, and they even have it in their slides with screenshots showing it worked at one point.
I tried a new project and observed the same behavior in a project with and without MasterDetailsPage this time. On Windows, the event fires when the back button from nav bar is pressed, however the breakpoint isn't hit. The event doesn't fire on Android. So same. So, it's inconsistent that the event fires on Windows but not other platforms. Maybe the ContentPage or Page documentation needs to be updated to mention this only occurs for the hardware back button except on Windows?
https://docs.microsoft.com/en-us/dotnet/api/Xamarin.Forms.ContentPage?view=xamarin-forms
https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.page?view=xamarin-forms
